filesystems/s3/src/main/java/org/lerch/s3fs/S3FileSystemProvider.java

```diff
@@ -426,19 +420,144 @@ public void copy(Path source, Path target, CopyOption... options) throws IOExcep
         }
 
-        String sourceBucketName = s3Source.getFileStore().name();
-        String keySource = s3Source.getKey();
-        String bucketNameTarget = s3Target.getFileStore().name();
-        String keyTarget = s3Target.getKey();
-        s3Source.getFileSystem()
-                .getClient()
-                .copyObject(CopyObjectRequest.builder()
-                        .copySource(sourceBucketName + "/" + keySource)
-                        .bucket(bucketNameTarget)
-                        .key(keyTarget)
-                        .build());
+        long objectSize = this.objectSize(s3Source);
+        long threshold = 5L * 1024L * 1024L * 1024L; //5GB
+        if (objectSize >=  threshold ){
+            // large file, do a multipart copy
+            multiPartCopy(s3Source, objectSize, s3Target, options);
+
+        } else {
+            //do a normal copy
+            String bucketNameOrigin = s3Source.getFileStore().name();
+            String keySource = s3Source.getKey();
+            String bucketNameTarget = s3Target.getFileStore().name();
+            String keyTarget = s3Target.getKey();
+            s3Source.getFileSystem()
+                    .getClient()
+                    .copyObject(CopyObjectRequest.builder()
+                            .copySource(bucketNameOrigin + "/" + keySource)
+                            .bucket(bucketNameTarget)
+                            .key(keyTarget)
+                            .build());
+        }
+    }
+
+    /**
+     * Copy large files
+     * https://docs.aws.amazon.com/AmazonS3/latest/API/API_UploadPartCopy.html
+     * @param source the object being copied
+     * @param target the destination
+     * @param options copy options
+     */
+    private void multiPartCopy(S3Path source, long objectSize, S3Path target, CopyOption... options) {
+        log.fine(() -> "Multipart copy initiated: source = " + source + ", objectSize = " + objectSize + ", target = " + target + ", options = " + Arrays.deepToString(options));
+        S3Client s3Client = target.getFileSystem().getClient();
+
+        final CreateMultipartUploadRequest createMultipartUploadRequest = CreateMultipartUploadRequest.builder()
+                .bucket(target.getFileStore().name())
+                .key(target.getKey())
+                .build();
+        final CreateMultipartUploadResponse createMultipartUploadResponse = s3Client.createMultipartUpload(createMultipartUploadRequest);
+        final String uploadId = createMultipartUploadResponse.uploadId();
+
+        // you can have at most 10K parts with at least one 5MB part
+        long partSize = Math.max((objectSize / 10000L )+1, 5 * 1024 * 1024);
+        long bytePosition = 0;
+        int partNum = 1;
+
+        List<CompletableFuture<UploadPartCopyResponse>> uploadFutures = new ArrayList<>();
+
+        // these are short running async I/O threads so we can have a lot of them
```

Comment:
Still looks a bit scary. Wouldn't it make sense to make this thread pool a class variable?

Reply:
Is the intent that the pool be a static variable and therefore effectively a singleton or do you want it to have a getter/setter

Reply:
I meant making it a static variable

Reply:
Done